blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+  const span = trace.getActiveSpan();
+  
+  try {
+    throw new Error('Something went wrong!');
+  } catch (error) {
+    if (span) {
+      span.setAttribute('error', true);
+      span.recordException(error);
+    }
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+}
+```
+This ensures that the error is captured within the OpenTelemetry trace and can be visualized in your tracing backend.
+
+### Exporting Metrics
+
+Next.js applications often benefit from metrics like request count, latency, and errors. Modify `otel.js`:
+
+```javascript
+import { MeterProvider } from '@opentelemetry/sdk-metrics';
+
+const meter = new MeterProvider().getMeter('nextjs-meter');
+const requestCounter = meter.createCounter('http_requests_total', {
+  description: 'Counts total HTTP requests',
+});
+
+export function trackRequest() {
+  requestCounter.add(1);
+}
+```
+
+Then, use it in an API route:
+```javascript
+import { trackRequest } from '../../otel';
+
+export default function handler(req, res) {
+  trackRequest();
+  res.status(200).json({ message: 'Metrics tracked!' });
+}
+```
+
+### Simplifying with @vercel/otel
+
+If you’re deploying your Next.js application on Vercel, the easiest way to integrate OpenTelemetry is by using the @vercel/otel package. This package simplifies the entire setup process and automatically enables distributed tracing for your application.
+
+What is @vercel/otel?
+
+@vercel/otel is an official package from Vercel that adds built-in support for OpenTelemetry in your Next.js app. It eliminates the need to manually configure OpenTelemetry SDKs, exporters, and instrumentations, offering you:
+- Automatic tracing for API routes, pages, and server-side rendering (SSR)
+- Built-in span creation for Next.js internals
+- Exception tracking out of the box
+- Export to OpenTelemetry-compatible backends (e.g., Highlight.io)
+
+We've covered this topic in more detail previously with our [blog on setting up @vercel/otel in next.js](./how-to-use-opentelemetry-to-monitor-nextjs-apps.md), so make sure to check it out for a full guide.
+
+### Putting it all together
+
+Let's put all of the pieces together and create a complete `otel.js` file that will automatically instrument your Next.js app. Using `@vercel/otel`, we'll configure export for Highlight.io, but you can use any other OpenTelemetry-compatible backend:
+
+```typescript
+import { registerOTel } from "@vercel/otel";
+import { SimpleSpanProcessor } from "@opentelemetry/sdk-trace-node";
+import { PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
+import { SimpleLogRecordProcessor } from "@opentelemetry/sdk-logs";
+import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
+import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-grpc";
+import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-grpc";
+
+export function register() {
+  if (process.env.NEXT_RUNTIME === "nodejs") {
```

Comment:
Is this conditional necessary? I thought the edge runtime was supported by @vercel/otel?